# Getaround project researchs

In [31]:
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.figure_factory as ff
import plotly.graph_objects as go
from plotly.subplots import make_subplots



## Load data

In [32]:
rawdata = pd.read_excel('../dashboard/src/data/get_around_delay_analysis.xlsx')
rawdata.head()

rental_id  car_id checkin_type  state  delay_at_checkout_in_minutes  \
0     575860  323777       mobile  ended                         -19.0   
1     571208  293910       mobile  ended                         -28.0   
2     561219  295860       mobile  ended                          -5.0   
3     575912  405611      connect  ended                         -58.0   
4     568455  382302      connect  ended                         -35.0   

   previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  
0                  575053.0                                        30.0  
1                  575013.0                                       600.0  
2                  574786.0                                        60.0  
3                  574676.0                                        90.0  
4                  574639.0                                       660.0

In [33]:
rawdata_pricing = pd.read_csv('../dashboard/src/data/get_around_pricing_project.csv')
rawdata_pricing.head()

Unnamed: 0 model_key  mileage  engine_power    fuel paint_color  \
0           0   Citroën   140411           100  diesel       black   
1           1   Citroën    13929           317  petrol        grey   
2           2   Citroën   183297           120  diesel       white   
3           3   Citroën   128035           135  diesel         red   
4           4   Citroën    97097           160  diesel      silver   

      car_type  private_parking_available  has_gps  has_air_conditioning  \
0  convertible                       True     True                 False   
1  convertible                       True     True                 False   
2  convertible                      False    False                 False   
3  convertible                       True     True                 False   
4  convertible                       True     True                 False   

   automatic_car  has_getaround_connect  has_speed_regulator  winter_tires  \
0          False                   True                 True          True   
1          False                  False                 True          True   
2          False                   True                False          True   
3          False                   True                 True          True   
4          False                  False                 True          True   

   rental_price_per_day  
0                   106  
1                   264  
2                   101  
3                   158  
4                   183

## Basics stats

In [34]:
# Basic stats
print("Taille du dataset:")
print("Number of rows : {}".format(rawdata.shape[0]))
print("Number of columns : {}".format(rawdata.shape[1]))
print()
print("---------------------------")
print()

print("Basics infos:")
print()
display(rawdata.info())
print()
print("---------------------------")
print()

print("Basics statistics: ")
print()
data_desc = rawdata.describe(include='all')
display(data_desc)
print()
print("---------------------------")
print()

print("Unique elements by feature: ")
print()
display(rawdata.nunique().sort_values())
print()
print("---------------------------")
print()

print("Percentage of missing values: ")
print()
display(100*rawdata.isnull().sum()/rawdata.shape[0])## 3. 

Taille du dataset:
Number of rows : 21310
Number of columns : 7

---------------------------

Basics infos:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21310 entries, 0 to 21309
Data columns (total 7 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   rental_id                                   21310 non-null  int64  
 1   car_id                                      21310 non-null  int64  
 2   checkin_type                                21310 non-null  object 
 3   state                                       21310 non-null  object 
 4   delay_at_checkout_in_minutes                16346 non-null  float64
 5   previous_ended_rental_id                    1841 non-null   float64
 6   time_delta_with_previous_rental_in_minutes  1841 non-null   float64
dtypes: float64(3), int64(2), object(2)
memory usage: 1.1+ MB


None


---------------------------

Basics statistics: 



rental_id         car_id checkin_type  state  \
count    21310.000000   21310.000000        21310  21310   
unique            NaN            NaN            2      2   
top               NaN            NaN       mobile  ended   
freq              NaN            NaN        17003  18045   
mean    549712.880338  350030.603426          NaN    NaN   
std      13863.446964   58206.249765          NaN    NaN   
min     504806.000000  159250.000000          NaN    NaN   
25%     540613.250000  317639.000000          NaN    NaN   
50%     550350.000000  368717.000000          NaN    NaN   
75%     560468.500000  394928.000000          NaN    NaN   
max     576401.000000  417675.000000          NaN    NaN   

        delay_at_checkout_in_minutes  previous_ended_rental_id  \
count                   16346.000000               1841.000000   
unique                           NaN                       NaN   
top                              NaN                       NaN   
freq                             NaN                       NaN   
mean                       59.701517             550127.411733   
std                      1002.561635              13184.023111   
min                    -22433.000000             505628.000000   
25%                       -36.000000             540896.000000   
50%                         9.000000             550567.000000   
75%                        67.000000             560823.000000   
max                     71084.000000             575053.000000   

        time_delta_with_previous_rental_in_minutes  
count                                  1841.000000  
unique                                         NaN  
top                                            NaN  
freq                                           NaN  
mean                                    279.288430  
std                                     254.594486  
min                                       0.000000  
25%                                      60.000000  
50%                                     180.000000  
75%                                     540.000000  
max                                     720.000000


---------------------------

Unique elements by feature: 



checkin_type                                      2
state                                             2
time_delta_with_previous_rental_in_minutes       25
delay_at_checkout_in_minutes                   1745
previous_ended_rental_id                       1788
car_id                                         8143
rental_id                                     21310
dtype: int64


---------------------------

Percentage of missing values: 



rental_id                                      0.000000
car_id                                         0.000000
checkin_type                                   0.000000
state                                          0.000000
delay_at_checkout_in_minutes                  23.294228
previous_ended_rental_id                      91.360863
time_delta_with_previous_rental_in_minutes    91.360863
dtype: float64

**Observations:**

Nous avons 2 données de type catégorielle :
- checkin_type: mobile(checkin en direct)/connect(checkin à distance)
- state: canceled/ended

et nous avons des données avec beaucoup de champs non renseignés :
- previous_ended_rental_id --> 91.36%
- time_delta_with_previous_rental_in_minutes --> 91.36%
- previous_ended_rental_id -- 23/29%

Il serait intéressant de:
- Voir le nombre location se faisant soit via le mobile soit directement.
- D'investiguer sur les données manquantes et prendre une décision sur comment les traiter.
- Voir s'il existe des valeurs abérantes pour les données : delay_at_checkout_in_minutes, time_delta_with_previous_rental_in_minutes

## EDA

In [35]:
data_pricing = rawdata_pricing[rawdata_pricing['rental_price_per_day'].notna()]

median_day_price = "%.2f" % data_pricing['rental_price_per_day'].median()
print(f"Loyer moyen journalier pour la location des voitures: {median_day_price}")

median_minute_price = "%.2f" % (float(median_day_price) / 60)
print(f"Loyer moyen par minute pour la location des voitures: {median_minute_price}")

Loyer moyen journalier pour la location des voitures: 119.00
Loyer moyen par minute pour la location des voitures: 1.98


In [36]:
dataset = rawdata.copy()

In [37]:
fig = px.pie(
    dataset,
    names='checkin_type',
    title='Répartition du type de récupération du véhicule'
)
fig.show()


In [38]:
fig = px.pie(
    dataset,
    names='state',
    title='Répartition des états des locations (finis ou annulées)'
)
fig.show()

Ajout d'une colonne permettant de voir quel était le délai de la dernière location annulée.
L'idee est de voir si il y a un lien entre le retard et l'annulation.

In [39]:
# Créer un dictionnaire pour mapper les previous_ended_rental_id à leurs delay_at_checkout_in_minutes correspondants
delay_map = dataset.set_index('rental_id')['delay_at_checkout_in_minutes'].to_dict()

# Utiliser la méthode apply avec une fonction lambda pour ajouter la nouvelle colonne
dataset['previous_ended_rental_delay_at_checkout'] = dataset['previous_ended_rental_id'].apply(lambda x: delay_map.get(x, None))
dataset[dataset["state"] == "canceled"].head(20)

dataset['loss_profit'] = dataset['delay_at_checkout_in_minutes'].apply(lambda x: 0 if x <= 0 else "%.2f" % (x * float(median_minute_price)))
dataset['loss_profit'] = dataset['loss_profit'].astype(float)


dataset.head(10)

rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
0     575860  323777       mobile     ended                         -19.0   
1     571208  293910       mobile     ended                         -28.0   
2     561219  295860       mobile     ended                          -5.0   
3     575912  405611      connect     ended                         -58.0   
4     568455  382302      connect     ended                         -35.0   
5     575497  356190       mobile     ended                         702.0   
6     571038  341861      connect     ended                        -152.0   
7     571552  403962      connect  canceled                           NaN   
8     575574  173061      connect     ended                          42.0   
9     572413  350149       mobile     ended                          79.0   

   previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  \
0                  575053.0                                        30.0   
1                  575013.0                                       600.0   
2                  574786.0                                        60.0   
3                  574676.0                                        90.0   
4                  574639.0                                       660.0   
5                  574596.0                                        30.0   
6                  574571.0                                       540.0   
7                  574540.0                                        90.0   
8                  574480.0                                       240.0   
9                  574325.0                                       600.0   

   previous_ended_rental_delay_at_checkout  loss_profit  
0                                     22.0         0.00  
1                                    -15.0         0.00  
2                                     65.0         0.00  
3                                     35.0         0.00  
4                                      6.0         0.00  
5                                     10.0      1389.96  
6                                    -54.0         0.00  
7                                    153.0          NaN  
8                                   -374.0        83.16  
9                                    169.0       156.42

In [40]:
ended_dataset = dataset[dataset['state'] == 'ended']
ended_dataset['loss_profit'].sum()

3757341.06

**Observation:**

Après l'ajout d'une colonne qui calcul la perte de bénéfice, nous pouvons obeserver que les différents retards pourraient engendrer dans le cas où le loueur loue sont véhicule juste après, une perte de **3.757.341$**.

---

Ajout d'une colonne représentant le nombre de fois qu'un véhicule a été loué.

In [41]:
dataset['rental_count'] = dataset.groupby('car_id')['car_id'].transform('count')
dataset.head()

rental_id  car_id checkin_type  state  delay_at_checkout_in_minutes  \
0     575860  323777       mobile  ended                         -19.0   
1     571208  293910       mobile  ended                         -28.0   
2     561219  295860       mobile  ended                          -5.0   
3     575912  405611      connect  ended                         -58.0   
4     568455  382302      connect  ended                         -35.0   

   previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  \
0                  575053.0                                        30.0   
1                  575013.0                                       600.0   
2                  574786.0                                        60.0   
3                  574676.0                                        90.0   
4                  574639.0                                       660.0   

   previous_ended_rental_delay_at_checkout  loss_profit  rental_count  
0                                     22.0          0.0             6  
1                                    -15.0          0.0            14  
2                                     65.0          0.0             6  
3                                     35.0          0.0             8  
4                                      6.0          0.0             8

On peut voir que la donnée previous_ended_rental_id n'est pas forcément renseigné.

In [42]:
test = dataset[dataset["car_id"] == 359049]
test.head(10)

rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
914       553735  359049      connect     ended                          33.0   
1467      537126  359049      connect     ended                        -156.0   
1843      508131  359049      connect     ended                          70.0   
7061      561550  359049      connect  canceled                           NaN   
7696      546894  359049      connect     ended                           2.0   
9385      544433  359049      connect     ended                        -110.0   
14316     539408  359049      connect     ended                          -6.0   
15179     563861  359049      connect  canceled                           NaN   
16381     547579  359049      connect     ended                          32.0   
17849     550645  359049      connect     ended                         249.0   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  \
914                    550645.0                                       390.0   
1467                   539408.0                                       720.0   
1843                        NaN                                         NaN   
7061                        NaN                                         NaN   
7696                        NaN                                         NaN   
9385                        NaN                                         NaN   
14316                       NaN                                         NaN   
15179                       NaN                                         NaN   
16381                       NaN                                         NaN   
17849                       NaN                                         NaN   

       previous_ended_rental_delay_at_checkout  loss_profit  rental_count  
914                                      249.0        65.34            10  
1467                                      -6.0         0.00            10  
1843                                       NaN       138.60            10  
7061                                       NaN          NaN            10  
7696                                       NaN         3.96            10  
9385                                       NaN         0.00            10  
14316                                      NaN         0.00            10  
15179                                      NaN          NaN            10  
16381                                      NaN        63.36            10  
17849                                      NaN       493.02            10

In [43]:
test2 = dataset[dataset["rental_id"] == 550645]
test2.head()

rental_id  car_id checkin_type  state  delay_at_checkout_in_minutes  \
17849     550645  359049      connect  ended                         249.0   

       previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  \
17849                       NaN                                         NaN   

       previous_ended_rental_delay_at_checkout  loss_profit  rental_count  
17849                                      NaN       493.02            10

In [44]:
# Visualisation de la distribution des données et repérage des outliers
fig = make_subplots(rows=1, cols=2, subplot_titles=("Délai d'attente pour rendre le véhicule en minutes", "Délai par rapport à la location précédente en minutes"))

fig.add_trace(go.Box(y=dataset['delay_at_checkout_in_minutes'], name='Delay'), row=1, col=1)
fig.add_trace(go.Box(y=dataset['time_delta_with_previous_rental_in_minutes'], name='Time delta'), row=1, col=2)

# Afficher la figure
fig.show()

Il y a des délais énormes sur la remise du véhicule. Il faudrait s'assurer que les données sont bonnes.

Il faudra surement se concentrer par la suite sur un échantillon plus restreint et ne pas prendre en compte les outliers.

In [45]:
# Show outliers
# Drop lines containing invalid values or outliers  [Xˉ−3σ,Xˉ+3σ][Xˉ−3σ,Xˉ+3σ]
def ouliers_viewver(dataset, sigmas=3, columns=[]):
    """
    Display outliers from Pandas dataset.

    Parameters:
    dataset (pd.DataFrame): Pandas dataset
    columns (list): list of the columns in dataset to check outliers. All by default. 
    
    Returns:
    Void
    """
    outliers_count = {}
    if len(columns) < 1:
        columns = dataset.columns
        
    for col in columns:
        mean = dataset[col].mean()
        std = dataset[col].std()
        
        # 3 sigmas rules
        lower_bound = mean - sigmas * std
        upper_bound = mean + sigmas * std

        #print(f"For col {col}, lower is {lower_bound} and upper is {upper_bound}")
        
        # Create mask
        outliers = (dataset[col] < lower_bound) | (dataset[col] > upper_bound)
        outliers_count[col] = outliers.sum()

    outliers_df = pd.DataFrame(list(outliers_count.items()), columns=['Column', 'Outliers'])
    display(outliers_df)


ouliers_viewver(dataset, sigmas=1, columns=["delay_at_checkout_in_minutes"])

Column  Outliers
0  delay_at_checkout_in_minutes       427

In [46]:
# Drop lines containing invalid values or outliers  [Xˉ−3σ,Xˉ+3σ][Xˉ−3σ,Xˉ+3σ]
def delete_ouliers(dataset, sigmas=3, columns=[]):
    """
    Delete outliers from Pandas dataset.

    1 sigma --> 68%
    2 sigmas --> 95%
    3 sigmas --> 99%


    Parameters:
    dataset (pd.DataFrame): Pandas dataset
    columns (list): list of the columns in dataset to check outliers. All by default. 
    
    Returns:
    pd.DataFrame: clean dataset
    """
    masks = []
    if len(columns) < 1:
        columns = dataset.columns
        
    for col in columns:
        mean = dataset[col].mean()
        std = dataset[col].std()
        
        # 3 sigmas rules
        lower_bound = mean - sigmas * std
        upper_bound = mean + sigmas * std
        #print(f"For col {col}, lower is {lower_bound} and upper is {upper_bound}")
        
        # Create mask
        mask = (dataset[col] >= lower_bound) & (dataset[col] <= upper_bound)
        masks.append(mask)

    # Apply mask in all columns
    # example: 
    # row1 = [0,1,1] -> [0]
    # row2 = [1,1,1] -> [1]
    final_mask = pd.concat(masks, axis=1).all(axis=1)
    filtered_df = dataset.loc[final_mask, :]
    return filtered_df


print("old dataset shape:", dataset.shape)
# dataset = delete_ouliers(dataset, columns=["delay_at_checkout_in_minutes"])
print("New dataset shape:", dataset.shape)

fig = px.box(
    dataset,
    y='delay_at_checkout_in_minutes',
    title='Répartition des délais pour rendre les véhicules en minute',
    labels={'delay_at_checkout_in_minutes': 'Minutes'}
)

# Afficher la figure
fig.show()

old dataset shape: (21310, 10)
New dataset shape: (21310, 10)


In [47]:
# Correlation matrix
corr_dataset = dataset[['rental_id','car_id','checkin_type','delay_at_checkout_in_minutes','previous_ended_rental_id','time_delta_with_previous_rental_in_minutes','rental_count']]
corr_dataset['checkin_type']  = corr_dataset['checkin_type'].apply(lambda x: 1 if x == 'connect' else 0)

correlation_matrix = corr_dataset.corr()

# Créer la figure
fig = go.Figure(data=go.Heatmap(
    z=correlation_matrix,
    x=correlation_matrix.columns,
    y=correlation_matrix.columns,
    zmin=-1,  # Minimum de l'échelle de couleurs
    zmax=1,   # Maximum de l'échelle de couleurs
    text=correlation_matrix.round(2),  # Valeurs à afficher dans les cellules
    texttemplate='%{text:.2f}',       # Format des valeurs
    textfont={"size": 10},            # Taille de la police pour les valeurs
    hoverongaps=False,                # Désactiver le survol sur les cellules vides
    colorscale='RdBu',                # Échelle de couleurs rouge-bleu
    colorbar=dict(
        title='Correlation',          # Titre de la barre de couleurs
        titleside='right',
        thickness=15
    )
))

# Mise à jour du layout
fig.update_layout(
    title='Matrice de Corrélation Interactive',
    width=900,                        # Largeur de la figure
    height=800,                       # Hauteur de la figure
    xaxis=dict(
        tickangle=45,                 # Rotation des labels de l'axe x
        side='bottom'
    ),
    yaxis=dict(
        autorange='reversed'          # Inverser l'axe y pour avoir la même disposition que seaborn
    )
)

# Afficher la figure
fig.show()

Il n'y a pas de fortes corrélations entres les données. Juste un peu entre le checking_type et le nombre de location pour un véhicule. Ce la est surement dû au fait qu'un utilisateur louant beaucoup son véhicule utilise un type de checkin (connect ou  mobile) et doit rester sur ce mode.

In [48]:
fig = make_subplots(rows=1, cols=2, subplot_titles=('Mobile', 'Connect'))

# Ajouter le premier graphique (histogramme)
fig.add_trace(
    go.Histogram(
        x=dataset[dataset['checkin_type'] == 'mobile']['delay_at_checkout_in_minutes'],
        xbins=dict( # recentrage
            start=-500,
            end=500,
            size=5
        ),
        name='Mobile',
        marker_color='blue',
    ),
    row=1, col=1
)

fig.add_trace(
    go.Histogram(
        x=dataset[dataset['checkin_type'] == 'connect']['delay_at_checkout_in_minutes'],
        xbins=dict( # recentrage
            start=-500,
            end=500,
            size=5
        ),
        name='Connect',
        marker_color='orange'
    ),
    row=1, col=2
)

fig.add_shape(
    type='line',
    x0=0, x1=0, 
    y0=0, y1=600,  
    line=dict(color='red', dash='dash'), 
    row=1, col=1  
)

fig.add_shape(
    type='line',
    x0=0, x1=0,  
    y0=0, y1=140,  
    line=dict(color='red', dash='dash'), 
    row=1, col=2  
)



fig.update_layout(
    title_text='Distribution des checkout par type d\'enregistrement', xaxis_title="Délais pour le checkout en minutes", yaxis_title="Nombre de checkout"
)

# Afficher le plot
fig.show()

In [49]:
def group_by_delay(x):
    if x < 0 :
        y = '0. Pas de retard'
    elif x < 15 : 
        y = '1. Retard < 15 min'
    elif x < 60 :
        y = '2. 15 ≤ Retard < 60 min'
    elif x >= 60 :
        y = '4. Retard ≥ 60 min'
    return y

dataset['delay'] = dataset['delay_at_checkout_in_minutes'].dropna().apply(lambda x: group_by_delay(x))

# Calculer le nombre total de chaque checkin_type
total_counts = dataset[dataset['delay'].notna()].groupby('checkin_type').size().reset_index(name='total_count')

delay_counts = dataset[dataset['delay'].notna()].groupby(['delay', 'checkin_type']).size().reset_index(name='count')

# Fusionner les DataFrames pour obtenir le total_count
delay_counts = delay_counts.merge(total_counts, on='checkin_type')

# Calculer le pourcentage de chaque catégorie de délai par checkin_type
delay_counts['percentage'] = (delay_counts['count'] / delay_counts['total_count']) * 100

# Créer le graphique à barres côte à côte
fig = px.bar(
    delay_counts,
    x='delay',
    y='percentage',
    color='checkin_type',
    barmode='group',
    title='Pourcentage des délais par type d\'enregistrement',
    labels={'delay': 'Catégorie de délai', 'percentage': 'Pourcentage', 'checkin_type': 'Type enregistrement'}
)

fig.update_traces(texttemplate='%{y:.2f}%', textposition='inside')

# Afficher le plot
fig.show()

In [50]:
fig = px.pie(
            dataset,
            names="delay",
            title="Distribution des États",
        )
fig.show()

**Observation:**

Nous pouvons observer qu'il y a plus de retards avec l'utilisation de l'enregistrement "mobile", c'est à dire en direct.

In [51]:
def group_by_delay(x):
    if x < 0 :
        y = '0. Pas de retard'
    elif x < 15 : 
        y = '1. Retard < 15 min'
    elif x < 60 :
        y = '2. 15 ≤ Retard < 60 min'
    elif x >= 60 :
        y = '4. Retard ≥ 60 min'
    return y

dataset['delay'] = dataset['delay_at_checkout_in_minutes'].dropna().apply(lambda x: group_by_delay(x))

# Calculer le nombre total de valeurs avec un delay
total_counts = dataset[dataset['delay'].notna()]['delay'].count()

delay_counts = dataset[dataset['delay'].notna()].groupby(['delay']).size().reset_index(name='count')


# # Calculer le pourcentage de chaque catégorie de délai par state
delay_counts['percentage'] = (delay_counts['count'] / total_counts) * 100

# # Créer le graphique à barres côte à côte
fig = px.bar(
    delay_counts,
    x='delay',
    y='percentage',
    color='percentage',
    barmode='group',
    title='Distribution des délais pour la récupération du véhicule',
    labels={'delay': 'Catégorie de délai', 'percentage': 'Pourcentage', 'checkin_type': 'Type enregistrement'}
)

fig.update_traces(texttemplate='%{y:.2f}%', textposition='inside')

# Afficher le plot
fig.show()

Pourquoi il y a t-il autant de délais plus important qu'une heure ?

**Observation:**

Nous pouvons observer qu'il y a plus de retards avec l'utilisation du mode direct.

In [52]:
dataset.head(20)

rental_id  car_id checkin_type     state  delay_at_checkout_in_minutes  \
0      575860  323777       mobile     ended                         -19.0   
1      571208  293910       mobile     ended                         -28.0   
2      561219  295860       mobile     ended                          -5.0   
3      575912  405611      connect     ended                         -58.0   
4      568455  382302      connect     ended                         -35.0   
5      575497  356190       mobile     ended                         702.0   
6      571038  341861      connect     ended                        -152.0   
7      571552  403962      connect  canceled                           NaN   
8      575574  173061      connect     ended                          42.0   
9      572413  350149       mobile     ended                          79.0   
10     573792  373511       mobile     ended                        -118.0   
11     571816  314220       mobile     ended                          60.0   
12     572707  241021      connect  canceled                           NaN   
13     560407  401387       mobile     ended                          43.0   
14     573770  356106      connect     ended                         -14.0   
15     568064  335276      connect     ended                         121.0   
16     576292  181625      connect     ended                           9.0   
17     572403  390467      connect     ended                           NaN   
18     574185  356540       mobile     ended                           NaN   
19     547428  160062       mobile  canceled                           NaN   

    previous_ended_rental_id  time_delta_with_previous_rental_in_minutes  \
0                   575053.0                                        30.0   
1                   575013.0                                       600.0   
2                   574786.0                                        60.0   
3                   574676.0                                        90.0   
4                   574639.0                                       660.0   
5                   574596.0                                        30.0   
6                   574571.0                                       540.0   
7                   574540.0                                        90.0   
8                   574480.0                                       240.0   
9                   574325.0                                       600.0   
10                  574297.0                                        30.0   
11                  574276.0                                       120.0   
12                  574151.0                                         0.0   
13                  574013.0                                        90.0   
14                  573913.0                                       390.0   
15                  573893.0                                       180.0   
16                  573698.0                                        90.0   
17                  573546.0                                       360.0   
18                  573533.0                                       180.0   
19                  573462.0                                       540.0   

    previous_ended_rental_delay_at_checkout  loss_profit  rental_count  \
0                                      22.0         0.00             6   
1                                     -15.0         0.00            14   
2                                      65.0         0.00             6   
3                                      35.0         0.00             8   
4                                       6.0         0.00             8   
5                                      10.0      1389.96             3   
6                                     -54.0         0.00            18   
7                                     153.0          NaN            13   
8                                    -374.0        83.16            13   
9                           

In [53]:
dataset.to_excel('get_around_eda.xlsx')

In [54]:
delay_range = range(0, 1000, 5)

# Calculer la perte totale pour chaque délai dans delay_range
loss_values = []
for delay in delay_range:
    adjusted_losses = dataset["delay_at_checkout_in_minutes"].apply(
        lambda x: (x - delay) * float(median_minute_price) if (x - delay) > 0 else 0
    )
    # Sum of losses
    total_loss = adjusted_losses.sum()
    loss_values.append(total_loss)

# Create a DataFrame for delay-dependent loss values
loss_data = pd.DataFrame(
    {
        "delay_threshold_between_rentals": delay_range,
        "total_loss": loss_values,
    }
)
fig = px.line(
    loss_data,
    x="delay_threshold_between_rentals",
    y="total_loss",
    title="Impact du seuil de retard sur la perte totale",
    labels={
        "delay_threshold_between_rentals": "Seuil de délai entre les locations (minutes)",
        "total_loss": "Perte total ($)",
    },
)

fig.show()

Visualisation des états de checkout en fonction du type de délai.

In [55]:
def categorize_rental_count(row):
    if row <= 10:
        return row
    else:
        return "Plus de 10"

# Appliquer la catégorisation
dataset["rental_category"] = dataset["rental_count"].apply(categorize_rental_count)

# Compter les occurrences de chaque état par catégorie
state_counts = (
    dataset.groupby(["state", "rental_category"]).size().reset_index(name="count")
)

# Créer le graphique en secteurs
fig = px.pie(
    state_counts,
    names="rental_category",
    values="count",
    title="Distribution des États par Catégorie de rental_count",
)
fig.show()

In [56]:
dataset["estimated_loss"] = dataset["delay_at_checkout_in_minutes"].apply(
    lambda x: (x - delay) * float(median_minute_price) if (x - delay) > 0 else 0
)
# Créer une colonne pour marquer les retards causant des pertes financières
dataset['is_loss_due_to_delay'] = dataset.apply(
    lambda x: x['delay_at_checkout_in_minutes'] > x['time_delta_with_previous_rental_in_minutes'],
    axis=1
)

# Calculer les pertes évitables pour différents seuils
results = []
for threshold in range(0, 800, 50):  # seuils de 0 à 600 minutes par pas de 50
    dataset['avoidable_with_threshold'] = (
        (dataset['time_delta_with_previous_rental_in_minutes'] < threshold) & 
        (dataset['is_loss_due_to_delay'])
    )
    # Compter le nombre de retards évitables et la perte financière évitable
    avoidable_count = dataset['avoidable_with_threshold'].sum()
    avoidable_loss = dataset.loc[dataset['avoidable_with_threshold'], 'estimated_loss'].sum()
    results.append({'Threshold': threshold, 'Avoidable Delays': avoidable_count, 'Avoidable Loss (€)': avoidable_loss})

# Convertir les résultats en DataFrame
result_df = pd.DataFrame(results)

# Créer le graphique en ligne
fig = px.line(result_df, x='Threshold', y='Avoidable Loss (€)', title="Impact financier des retards en fonction du seuil entre locations",
              labels={'Threshold': 'Seuil minimum entre locations (minutes)', 'Avoidable Loss (€)': 'Pertes évitables (€)'})

fig.show()


In [57]:
dataset['critical_delay_for_next_rental'] = dataset['time_delta_with_previous_rental_in_minutes'] - dataset['previous_ended_rental_delay_at_checkout']
canceled_dataset = dataset[dataset['state'] == 'canceled']
dataset['is_cancel_due_to_delay_by_previous_rental'] = canceled_dataset['critical_delay_for_next_rental'].apply(lambda x: True if x < 0 else False)



In [58]:
# Filtrer les locations ayant une valeur dans time_delta_with_previous_rental_in_minutes
filtered_data = dataset[dataset["time_delta_with_previous_rental_in_minutes"].notnull()]

# Créer une colonne pour déterminer si la location est en retard ou non
filtered_data["is_late"] = filtered_data["delay_at_checkout_in_minutes"] > 0

# Graphique 1 : Pourcentage de locations en retard
late_counts = filtered_data["is_late"].value_counts()
fig = px.pie(
    names=late_counts.index.map({True: "En retard", False: "À l'heure"}),
    values=late_counts.values,
    title="Pourcentage de locations en retard parmi celles avec une location précédente"
)
fig.show()

In [59]:
# Graphique 2 : Pourcentage de locations annulées en raison d'un retard
cancelled_counts = filtered_data["is_cancel_due_to_delay_by_previous_rental"].value_counts()
fig = px.pie(
    names=cancelled_counts.index.map({True: "En retard et annulé", False: "À l'heure et annulé"}),
    values=cancelled_counts.values,
    title="Pourcentage de locations en retard entraînant une annulation"
)

# Affichage des graphiques
fig.show()

In [60]:
def prepare_data(data):
    delay_map = data.set_index("rental_id")["delay_at_checkout_in_minutes"].to_dict()
    data["previous_ended_rental_delay_at_checkout"] = data[
        "previous_ended_rental_id"
    ].apply(lambda x: delay_map.get(x, None))

    data[data["state"] == "canceled"].head(20)

    data["profit"] = data["delay_at_checkout_in_minutes"].apply(
        lambda x: "%.2f" % (x * float(median_minute_price))
    )

    data["profit"] = data["profit"].astype(float)

    data["delay"] = (
        data["delay_at_checkout_in_minutes"].dropna().apply(lambda x: group_by_delay(x))
    )

    data["rental_count"] = data.groupby("car_id")["car_id"].transform("count")

    data["critical_delay_for_next_rental_in_minutes"] = (
        data["time_delta_with_previous_rental_in_minutes"]
        - data["previous_ended_rental_delay_at_checkout"]
    )

    # Create a column to mark delays causing potential financial losses
    filtered_dataset = data[(data["delay_at_checkout_in_minutes"].notna())]
    data["is_potential_loss_due_to_delay"] = data["delay_at_checkout_in_minutes"].apply(
        lambda x: True if x > 0 else False,
    )

    filtered_dataset = data[(data["previous_ended_rental_delay_at_checkout"].notna())]
    data["is_cancel_due_to_delay_by_previous_rental"] = filtered_dataset[
        "critical_delay_for_next_rental_in_minutes"
    ].apply(lambda x: True if x < 0 else False)

    return data

dataprepared = prepare_data(data=rawdata)
# data = delete_ouliers(
#     dataset=dataprepared, sigmas=3, columns=["delay_at_checkout_in_minutes"]
# )
dataprepared.to_excel("delay_analyse_clean.xlsx")